pkg/sql/alter_column_type_test.go

```diff
@@ -0,0 +1,223 @@
+// Copyright 2020 The Cockroach Authors.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package sql_test
+
+import (
+	"context"
+	"sync"
+	"testing"
+
+	"github.com/cockroachdb/cockroach/pkg/base"
+	"github.com/cockroachdb/cockroach/pkg/sql"
+	"github.com/cockroachdb/cockroach/pkg/sql/sqlbase"
+	"github.com/cockroachdb/cockroach/pkg/sql/tests"
+	"github.com/cockroachdb/cockroach/pkg/testutils/serverutils"
+	"github.com/cockroachdb/cockroach/pkg/util/leaktest"
+)
+
+func TestAlterColumnTypeInTxn(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+	ctx := context.Background()
+	params, _ := tests.CreateTestServerParams()
+	s, sqlDB, kvDB := serverutils.StartServer(t, params)
+	defer s.Stopper().Stop(ctx)
+
+	if _, err := sqlDB.Exec(`
+CREATE DATABASE t;
+BEGIN;
+CREATE TABLE t.test (x INT);
+ALTER TABLE t.test ALTER COLUMN x TYPE STRING;
+INSERT INTO t.test VALUES ('hello');
+COMMIT;
+`); err != nil {
+		t.Fatal(err)
+	}
+	// Ensure that t.test doesn't have any pending mutations
+	// after the AlterColumnType change.
+	desc := sqlbase.GetTableDescriptor(kvDB, "t", "test")
+	if len(desc.Mutations) != 0 {
+		t.Fatalf("expected to find 0 mutations, but found %d", len(desc.Mutations))
+	}
+}
+
+// Test while old column still exists - parse error.
+func TestInsertBeforeOldColumnIsDropped(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+
+	params, _ := tests.CreateTestServerParams()
+
+	blockGC := make(chan struct{})
+	params.Knobs = base.TestingKnobs{
+		GCJob: &sql.GCJobTestingKnobs{RunBeforeResume: func(_ int64) error { <-blockGC; return nil }},
```

Comment:
I don't think the GC job is the right one to block -- you want to block the schema change job adopter.